### ISSUE TITLE
Wanted: accurate preview

### ISSUE THREAD

**Issue by rtollert** [ni/labview-icon-editor#8]
You can go for a long time as an advanced LabVIEW user without realizing that while the LabVIEW Icon _Editor_ is 24-bit, LabVIEW *icons* are 8-bit (the [web-safe](https://en.wikipedia.org/wiki/Web_colors#Web-safe_colors) colors, to be precise, along with some extra grays). For instance, these gradients:
![2024-05-24-114256_68x67_scrot](https://github.com/ni/labview-icon-editor/assets/5639882/7af22152-586c-4c35-bec1-7830beb3fbb1)
are crushed into:
![2024-05-24-114307_38x43_scrot](https://github.com/ni/labview-icon-editor/assets/5639882/ee077759-312c-4b1a-a66c-db8d8fcc64bc)
And you have *no feedback* in the Icon Editor that this is about to happen, until you click OK. (Fortunately, the Icon Editor data model is also 24-bit, so you can open it back up to fix any issues that arise.)

The most annoying situation for this sort of thing is when you want to apply a slight tint to white, or perhaps some shade of gray. You're only going to be *specifically* doing that (and caring about the result) with an area fill. And that's precisely where any color differences will be most notable. Perhaps the worst-case scenario is in the light violets, with significantly more blue than red, in order to maximize the misalignment with the 6x6x6 color cube of the web-safe palette: #f7efff is crushed to white, while the only slightly-darker #f2e5ff goes off in a completely-different-yet-equally-unacceptable direction.
![2024-05-24-125128_67x64_scrot](https://github.com/ni/labview-icon-editor/assets/5639882/0a906bff-6596-418f-ac2d-6c57c0d3d475)
![2024-05-24-125134_43x47_scrot](https://github.com/ni/labview-icon-editor/assets/5639882/060fa2ab-b096-4630-aaa9-ab58689df165)

This can also lead to trouble when you want to apply some sort of overlay with a light alpha: at the high transparency settings you tend to want, only some of the pixels will actually be any different.

Ways to potentially solve this:
- The icon preview in the Icon Editor should render the icon the same way LabVIEW would, including the color depth conversion.
- Ask NI to add 32-bit icon support.
- (Maybe) some better dither support, but I don't know how I feel about this, my experience with dither has been generally negative

**Comment by JayKayAce** [comment]
An alternative way would be to limit the color pallette in the Icon editor to the websafe colors. This would be less intrusive to the LabVIEW VI Format, until the VI Format supports 24bit Icons.

I have cursed this feature multiple times because I was drawing some iconography for VIs using the TestStand green, which then was compressed into another green, that did not look anything like the amount of work I had put in for finding the right green.

**Comment by crossrulz** [comment]
It would also be useful to have a viewer that shows what the icon will look like when it is on a block diagram, including size (not zoomed in)

**Comment by JayKayAce** [comment]
Besides this preview box
![image](https://github.com/user-attachments/assets/3c1f4b58-9c11-43c3-85e8-5d6c119a65aa)

**Comment by crossrulz** [comment]
Good catch.  That's what I get for not double checking before making a post...

**Comment by crossrulz** [comment]
How is the icon stored in a VI file? Is it saved as an 8-bit, web-safe image? Or is it a different format (24 or 32-bit images)?

If the icon is saved as an 8-bit image, then it makes sense for the editor to limit the color palette to compatible colors. If not, then it is recommended to see if the block diagram renderer handle the larger color palette.

**Comment by svelderrainruiz** [comment]
> How is the icon stored in a VI file? Is it saved as an 8-bit, web-safe image? Or is it a different format (24 or 32-bit images)?
> 
> If the icon is saved as an 8-bit image, then it makes sense for the editor to limit the color palette to compatible colors. If not, then it is recommended to see if the block diagram renderer handle the larger color palette.

@gregr-ni  Can you please help us understand this?

**Comment by mballawickersigpi** [comment]
![Image](https://github.com/user-attachments/assets/9f78e901-e41d-4b63-9376-5923c66191c6) I have a color picker vi used in my custom IE. It will only show the acceptable icon colors used by icons. The color picker is called via mouse up event on a color box indicator. Because the color box is an indicator the windows/OS color picker is not called so there are no calling color picker conflicts.

**Comment by rtollert** [comment]
> How is the icon stored in a VI file? Is it saved as an 8-bit, web-safe image? Or is it a different format (24 or 32-bit images)?

The 8-bit icon is stored as a resource, but the (24-bit) Icon Editor data is stored somewhere else in the VI. The palette includes the web-safe colors, but it also includes 4-bit ramps for greyscale, red, green, and blue. The exact palette is returned out of **VI:VI Icon.Get as Image Data**. Deducting duplicates, 6³ + 4*(2⁴ - 6) = 256.

> If the icon is saved as an 8-bit image, then it makes sense for the editor to limit the color palette to compatible colors. If not, then it is recommended to see if the block diagram renderer handle the larger color palette.

Yes, those are two of the potential solutions I outlined in the OP. I think option \#1 *might* be as simple as bouncing the 32x32 preview through **VI:VI Icon.Set from Image Data** and **VI:VI Icon.Get as Image Data** (against some dummy VI).

**Comment by mballawickersigpi** [comment]
Here is a LAVAg post back from 2007 where Justin actually did some testing showing that the IE color picker only allows 221 colors. https://lavag.org/topic/4638-labviews-256-color-vi-icon-colormap/

**Comment by mballawickersigpi** [comment]
Wrote a vi in LV 2023 that was able to reproduce the issue showing that the color picker changes a (0,0,136) color to a (0,0,153)
even though the 0,0,136 is in the 8 bit color pallet.

![Image](https://github.com/user-attachments/assets/3e466e22-6c0b-4576-92e2-c66bbbf8837d)

**Comment by mballawickersigpi** [comment]
Additional testing showed that 221 of the 256 colors are valid Vi icon colors.

![Image](https://github.com/user-attachments/assets/b461c908-1d98-4f27-bcb0-2b90cdd110eb)

**Comment by mballawickersigpi** [comment]
To address the saving of the 32 bit colors the ni IE bundles up all the IE data into a cluster.

![Image](https://github.com/user-attachments/assets/820d73bf-e059-412e-8acb-4d0d3d0ccbd8)

It then send the flatten cluster to the lv_icon.lvlib:Write Icon Data to VI. which is locked.

![Image](https://github.com/user-attachments/assets/e6cdff25-fe7b-4a8d-b717-ac99102b55ad)

The lv_icon.lvlib:Write Icon Data to Library.vi however is not locked and it uses the set tag function to store the flatten cluster in the VI/Library data so I assume the locked write icon data to VI does it the same way.

![Image](https://github.com/user-attachments/assets/98c19667-e190-40ba-a279-a6861f99bfca)

So the 32 bit colors are only used in the ni IE UI and not related to how the VI is shown on the block diagram.

**Comment by svelderrainruiz** [comment]
We will split this issue into 2. We need to get further guidance on how 32 bit colors can be supported on the block diagram.

**Comment by svelderrainruiz** [comment]
Will move to a discussion after talking about it on todays steerco meeting
